packages/go_router/example/lib/stateful_shell_route.dart

```diff
@@ -0,0 +1,550 @@
+                          const TabScreen(
+                              label: 'C1', detailsPath: '/c1/details'),
+                      routes: <RouteBase>[
+                        GoRoute(
+                          path: 'details',
+                          builder:
+                              (BuildContext context, GoRouterState state) =>
+                                  DetailsScreen(
+                            label: 'C1',
+                            extra: state.extra,
+                            withScaffold: false,
+                          ),
+                        ),
+                      ],
+                    ),
+                  ]),
+                  StatefulShellBranch(routes: <GoRoute>[
+                    GoRoute(
+                      path: '/c2',
+                      builder: (BuildContext context, GoRouterState state) =>
+                          const TabScreen(
+                              label: 'C2', detailsPath: '/c2/details'),
+                      routes: <RouteBase>[
+                        GoRoute(
+                          path: 'details',
+                          builder:
+                              (BuildContext context, GoRouterState state) =>
+                                  DetailsScreen(
+                            label: 'C2',
+                            extra: state.extra,
+                            withScaffold: false,
+                          ),
+                        ),
+                      ],
+                    ),
+                  ]),
+                ],
+                builder:
+                    (BuildContext context, GoRouterState state, Widget child) {
+                  /// For this nested StatefulShellRoute we are using a custom
+                  /// container (TabBarView) for the branch navigators, and thus
+                  /// ignoring the default navigator contained passed to the
+                  /// builder. Custom implementation can access the branch
+                  /// navigators via the StatefulShellRouteState
+                  /// (see TabbedRootScreen for details).
+                  return const TabbedRootScreen();
+                },
+              ),
+            ],
+          ),
+        ],
+        builder: (BuildContext context, GoRouterState state, Widget child) {
+          return ScaffoldWithNavBar(body: child);
+        },
+
+        /// It's possible to customize the container for the branch navigators
+        /// even further, to for instance setup custom animations. The code
+        /// below is an example of such a customization (see
+        /// _AnimatedRouteBranchContainer). Note that in this case, the Widget
+        /// provided in the child parameter should not be added to the widget
+        /// tree. Instead, access the child widgets of each branch directly
+        /// (see StatefulShellRouteState.children) to implement a custom layout
```

Comment:
Yes, well, I can agree accessing the branch widgets/navigators via the state might not be most optimal way, but I think I only introduced this to enable using the same builder function signature as `ShellRoute`, and also due to feedback with suggestions around introducing a state class. However I now think a better general API is achieved by using other inputs to the builder functions (see below). 

The only reason the `children` field exists is to be able to support building a custom handling for the branch Navigators, instead of using the default `IndexedStack` solution. However I believe there are better ways than using `StatefulShellRouteState`. I've implemented a different solution in the https://github.com/tolo/flutter_packages/tree/nested-persistent-navigation-alternative branch, where both the builder and pageBuilder contains a single `StatefulShellBuilder` parameter, that provides support for building the shell along with its body. This makes it possible to specify **_either_** a builder or a pageBuilder, much like `ShellRoute` and `GoRoute`. When using `StatefulShellBuilder` to build the shell, you pass a callback for building the body of the shell. The last argument to this callback is a child Widget (`ShellNavigatorContainer`, internally using an `IndexedStack`), which exposes a `List<Widget> children` field. See below examples:

Normal use, i.e. using the default `IndexedStack` provided by `ShellNavigatorContainer` (`_IndexedStackedRouteBranchContainer`): 
```dart
builder: (StatefulShellBuilder shellBuilder) {
  return shellBuilder.buildShell((BuildContext context,
          GoRouterState state, ShellNavigatorContainer child) =>
      ScaffoldWithNavBar(body: child));
}
```

Normal use when instead using the pageBuilder looks like this: 
```dart
pageBuilder: (StatefulShellBuilder shellBuilder) {
  final Widget shell = shellBuilder.buildShell((BuildContext context, 
          GoRouterState state, ShellNavigatorContainer child) =>
      ScaffoldWithNavBar(body: child));
  return NoTransitionPage<dynamic>(child: shell);
}
```

Customised use, accessing the children, i.e. branch Navigators, directly, and for instance placing them in a TabBarView, PageView, Stack etc: 
```dart
builder: (StatefulShellBuilder shellBuilder) {
  return shellBuilder.buildShell(
    (BuildContext context, GoRouterState state,
            ShellNavigatorContainer child) =>
        TabbedRootScreen(children: child.children),
  );
}
```

Reply:
When building the body why does it have to be wrapped in `StatefulNavigationShell` ? It looks like the only difference is that the context of InheritedStatefulNavigationShell? which the main purposes is to host StatefulShellRouteState?

If so, can the StatefulShellRouteState be passed into  both builder and pageBuilder? that way you can get rid of  StatefulShellBuilder. The StatefulShellBuilder seems a bit redundant to me

Reply:
I guess part of the "problem" here is that `StatefulNavigationShell` needs to be above the custom shell body widget, since it provides and manages the `StatefulShellRouteState`. So the builders needs to return the `StatefulNavigationShell` as the Widget / Page child, without necessarily having to deal directly with the creation of it. So the crux here really is how to accomplish this in as easy a way as possible. Using `StatefulShellBuilder` at least abstracts away the creation of `StatefulNavigationShell`. 

Below is a diagram trying to explain how the parts fit together: 
![StatefulShellRoute-architecture](https://user-images.githubusercontent.com/1228601/219873628-fb553c89-3ea8-4055-b4d4-4d91237bcad2.svg)